apps/aecore/src/pow_sha256.erl

```diff
@@ -0,0 +1,109 @@
+%%% @end
+%%%=============================================================================
+-module(pow_sha256).
+
+-export([generate/3,
+         recalculate_difficulty/3,
+         pick_nonce/0]).
+
+-ifdef(TEST).
+-compile(export_all).
+-endif.
+
+-include("sha256.hrl").
+
+%% 10^24, approx. 2^80
+-define(NONCE_RANGE, 1000000000000000000000000).
+
+%% Integers represented in scientific notation:
+%%   256 * <base-2 exponent> + the most significant byte (i.e., the significand,
+%%   see https://en.wikipedia.org/wiki/Significand)
+-type sci_int() :: integer().
+-type pow_result() :: {ok, integer()} | {'error', 'generation_count_exhausted'}.
+
+-export_type([sci_int/0,
+              pow_result/0]).
+
+%%%=============================================================================
+%%% API
+%%%=============================================================================
+
+%%------------------------------------------------------------------------------
+%% Generate a nonce value that, when used in hash calculation of 'Data', results in
+%% a smaller value than the difficulty. Return {error, not_found} if the condition
+%% is still not satisfied after trying 'Count' times with incremented nonce values.
+%%------------------------------------------------------------------------------
+-spec generate(Data :: sha256:hashable(), Difficulty :: sci_int(), Count :: integer()) ->
+                      pow_result().
+generate(Data, Difficulty, Count) ->
+    %% Pick a nonce between 0 and 10^24 (approx. 2^80)
+    Nonce = ?MODULE:pick_nonce(),
+    generate_with_nonce(Data, Difficulty, Nonce, Count).
+
+%%------------------------------------------------------------------------------
+%% Adjust difficulty so that generation of new blocks proceeds at the expected pace
+%%------------------------------------------------------------------------------
+-spec recalculate_difficulty(sci_int(), integer(), integer()) -> sci_int().
+recalculate_difficulty(Difficulty, Expected, Actual) ->
+    DiffInt = scientific_to_integer(Difficulty),
+    integer_to_scientific(max(1, DiffInt * Expected div Actual)).
+
+
+%%%=============================================================================
+%%% Internal functions
+%%%=============================================================================
+
+
+-spec generate_with_nonce(Data :: sha256:hashable(), Difficulty :: sci_int(),
+                          Nonce :: integer(), Count :: integer()) -> pow_result().
+generate_with_nonce(_Data, _Difficulty, _Nonce, 0) ->
+    %% Count exhausted: fail
+    {error, generation_count_exhausted};
+generate_with_nonce(Data, Difficulty, Nonce, Count) ->
+    Hash1 = sha256:hash(Data),
+    Hash2 = sha256:hash(<<Hash1/binary, Difficulty:16, Nonce:?HASH_BITS>>),
+    case binary_to_scientific(Hash2) of
+        HD when HD < Difficulty ->
+            %% Hash satisfies condition: return nonce
+            {ok, Nonce};
```

Comment:
In principle the nonce returned may not be the same nonce serialized in the hash. May you please add a comment if there is an assumption here?

The nonce is initialized to a number - let us assume the max e.g. `Nonce = ((10^24) - 1)`. At each attempt (see `Count - 1`) the nonce is increased (see `Nonce + 1`). For the sake of hash computation, the nonce used is the 256 least significant bits of it (see `Nonce:?HASH_BITS`; e.g. `<<257:8>>` evaluates to 
`<<1>>`).